cozy/syntax.py

```diff
@@ -73,6 +73,10 @@ class BOp(object):
 
 class Exp(ADT):
     def with_type(self, t):
+        if hasattr(self, "type"):
+            if self.type == t:
+                return self
+            raise Exception("{} -> {}".format(self.type, t))
```

Comment:
I would like this to have a better error message. We can also refine the exception type to [`ValueError`](https://docs.python.org/3/library/exceptions.html#ValueError):

> Raised when an operation or function receives an argument that has the right type but an inappropriate value

```suggestion
            raise ValueError("The type of expression {} was changed from {} to {}. The type of an expression should only be assigned once!".format(self, self.type, t))
```